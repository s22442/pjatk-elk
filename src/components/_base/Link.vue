<script setup lang="ts">
interface Props {
  to?: string;
  external?: boolean;
  replace?: boolean;
}

const { to, external = false, replace = false } = defineProps<Props>();

const router = useRouter();

const handleClick = () => {
  if (!to) {
    return;
  }

  if (external) {
    window.open(to, '_blank')?.focus();
  } else if (replace) {
    router.replace(to);
  } else {
    router.push(to);
  }
};
</script>

<template>
  <a
    :href="to || 'javascript:;'"
    _inline-block
    _underline
    _fw500
    _text-secondary-interactive
    _transition-colors
    _cursor-pointer
    _select-none
    @click.prevent="handleClick()"
  >
    <slot />
  </a>
</template>
