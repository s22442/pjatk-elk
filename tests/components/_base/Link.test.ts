import Link from '~/components/_base/Link.vue';

import { mount } from '@vue/test-utils';

describe('Link', () => {
  it('should render', () => {
    const wrapper = mount(Link, { slots: { default: 'abc' } });
    expect(wrapper.text()).toContain('abc');
    expect(wrapper.html()).toMatchSnapshot();
  });

  const to = '/abc';

  const openSpy = vi.spyOn(window, 'open').mockImplementation(() => window);
  const focusSpy = vi.spyOn(window, 'focus').mockImplementation(() => {});

  it('should do nothing', async () => {
    const wrapper = mount(Link);

    await wrapper.trigger('click');

    expect(wrapper.router.push).not.toHaveBeenCalled();
    expect(wrapper.router.replace).not.toHaveBeenCalled();
    expect(openSpy).not.toHaveBeenCalled();
    expect(focusSpy).not.toHaveBeenCalled();
  });

  it('should open external link', async () => {
    const wrapper = mount(Link, { props: { to, external: true } });

    await wrapper.trigger('click');

    expect(openSpy).toHaveBeenCalledOnce();
    expect(openSpy).toHaveBeenCalledWith(to, '_blank');
    expect(focusSpy).toHaveBeenCalledOnce();
  });

  it('should push router', async () => {
    const wrapper = mount(Link, { props: { to } });

    await wrapper.trigger('click');

    expect(wrapper.router.push).toHaveBeenCalledOnce();
    expect(wrapper.router.push).toHaveBeenCalledWith(to);
  });

  it('should replace router', async () => {
    const wrapper = mount(Link, { props: { to, replace: true } });

    await wrapper.trigger('click');

    expect(wrapper.router.replace).toHaveBeenCalledOnce();
    expect(wrapper.router.replace).toHaveBeenCalledWith(to);
  });
});
