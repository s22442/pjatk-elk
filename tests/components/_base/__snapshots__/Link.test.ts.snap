// Vitest Snapshot v1

exports[`Link > should render 1`] = `"<a href=\\"javascript:;\\" _inline-block=\\"\\" _underline=\\"\\" _fw500=\\"\\" _text-secondary-interactive=\\"\\" _transition-colors=\\"\\" _cursor-pointer=\\"\\" _select-none=\\"\\">abc</a>"`;
